packages/venia-concept/src/components/ProductImageCarousel/carousel.js

```diff
@@ -1,138 +1,114 @@
 
-    rightChevronHandler = () => {
-        const sortedImages = this.sortedImages;
-        const { activeItemIndex } = this.state;
-        this.updateActiveItemIndex((activeItemIndex + 1) % sortedImages.length);
-    };
+    const [carouselState, carouselApi] = useCarousel(props.images);
+    const { activeItemIndex, sortedImages } = carouselState;
+    const { handlePrevious, handleNext, setActiveItemIndex } = carouselApi;
 
-    getChevron = direction => (
-        <button
-            onClick={this[`${direction}ChevronHandler`]}
-            className={this.props.classes[`chevron-${direction}`]}
-        >
-            <Icon src={ChevronIcons[direction]} size={40} />
-        </button>
+    const handleThumbnailClick = useCallback(
+        index => {
+            setActiveItemIndex(index);
+        },
+        [setActiveItemIndex]
     );
 
-    setCurrentImageLoaded = () => {
-        this.setState({
-            currentImageLoaded: true
-        });
-    };
+    const currentImage = sortedImages[activeItemIndex] || {};
 
-    render() {
-        const { classes } = this.props;
+    const src = currentImage.file
+        ? resourceUrl(currentImage.file, {
+              type: 'image-product',
+              width: DEFAULT_IMAGE_WIDTH
+          })
+        : transparentPlaceholder;
 
-        const sortedImages = this.sortedImages;
+    const alt = currentImage.label || 'image-product';
 
-        const mainImage = sortedImages[this.state.activeItemIndex] || {};
-        const src = mainImage.file
-            ? resourceUrl(mainImage.file, { type: 'image-product', width: 640 })
-            : transparentPlaceholder;
-        const alt = mainImage.label || 'image-product';
-        return (
-            <div className={classes.root}>
-                <div className={classes.imageContainer}>
-                    {this.getChevron('left')}
-                    <img
-                        onLoad={this.setCurrentImageLoaded}
-                        className={classes.currentImage}
-                        src={src}
-                        alt={alt}
-                    />
-                    {!this.state.currentImageLoaded ? (
-                        <img
-                            className={classes.currentImage}
-                            src={transparentPlaceholder}
-                            alt={alt}
-                        />
-                    ) : null}
-                    {this.getChevron('right')}
-                </div>
-                <ThumbnailList
-                    items={sortedImages}
-                    activeItemIndex={this.state.activeItemIndex}
-                    updateActiveItemIndex={this.updateActiveItemIndex}
+    const thumbnails = useMemo(
+        () =>
+            sortedImages.map((item, index) => (
+                <Thumbnail
+                    key={index}
+                    item={item}
+                    itemIndex={index}
+                    isActive={activeItemIndex === index}
```

Comment:
Is there a better way to do this? As is we will currently re-calculate this map every time we click next/back since `activeItemIndex` will update.

Reply:
What am I saying, we key them and react should handle the re-rendering based on the prop changes...

Reply:
Not really; a little mapping like this is not a big deal, and an optimization would probably make it less obvious what's happening here. But `key` should probably be `item.file` instead of `index`.